<font size="7"><p style="text-align: center;">**Gérer ses données avec python**</font></p>
<br><br><br>
<img src="img\pandas.svg">

## <u>1. Importer des données avec Pandas</u>
Pandas offre une panoplie de connecteurs pour importer des données (CSV, Excel, SQL, hdfs, etc.).

data source: https://www.kaggle.com/rashikrahmanpritom/heart-attack-analysis-prediction-dataset

In [ ]:
!pip install pyodbc

In [1]:
!pip install lxml

In [2]:
!pip install pandas

In [99]:
!pip install sqlalchemy

In [4]:
import numpy as np
import pandas as pd
import copy as cp
import matplotlib.pyplot as plt
import requests as rq
from io import StringIO


In [17]:
data = pd.read_csv('heart.csv')

# La méthode head() permet de n'afficher que les premières lignes (5 par défaut)
# Ici on en affiche 10
display(data.head(10))
display(data.tail())

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   
5   57    1   0     140   192    0        1       148     0      0.4    1   
6   56    0   1     140   294    0        0       153     0      1.3    1   
7   44    1   1     120   263    0        1       173     0      0.0    2   
8   52    1   2     172   199    1        1       162     0      0.5    2   
9   57    1   2     150   168    0        1       174     0      1.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0

In [18]:
test = data[['age','sex','cp']]
test.to_csv('test.csv')


Il existe 2 structures de donnée dans pandas:
- les **Series**: correspond à une colonne, à un numpy array avec un type de donnée unique
- les **DataFrame**: correspond à une collection de **Series**

## <u>2. Comment manipuler ses données avec pandas ?</u>

**2.1. Indexation et colonnes**

In [7]:
data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [19]:
data[['cp','age']]

cp  age
0     3   63
1     2   37
2     1   41
3     1   56
4     0   57
..   ..  ...
298   0   57
299   3   45
300   0   68
301   0   57
302   1   57

[303 rows x 2 columns]

In [ ]:
data.cp

0      3
1      2
2      1
3      1
4      0
      ..
298    0
299    3
300    0
301    0
302    1
Name: cp, Length: 303, dtype: int64

In [11]:
test = pd.DataFrame({'nom':['Jules','Thomas','Marc'],
                    'age':[40,50,60]},index=['a','b','c'])
test

nom  age
a   Jules   40
b  Thomas   50
c    Marc   60

In [13]:
test.loc['a':'c',['nom']]

nom
a   Jules
b  Thomas
c    Marc

In [17]:
test.iloc[0,0]

'Jules'

In [10]:
data.loc[5:10,['sex','cp']]

sex  cp
5     1   0
6     0   1
7     1   1
8     1   2
9     1   2
10    1   0

In [14]:
data.iloc[5:10,0:5]

age  sex  cp  trtbps  chol
5   57    1   0     140   192
6   56    0   1     140   294
7   44    1   1     120   263
8   52    1   2     172   199
9   57    1   2     150   168

In [23]:
data.at[0,'age']

np.int64(63)

In [24]:
data.iat[0,0]

np.int64(63)

In [17]:
# Avec le slicing, comme pour les listes python
# En ne sélectionnant qu'une seule ligne
data_sliced = data.loc[[10]]
display(data_sliced)
# Selectionner plusieurs lignes avec une list ou une array
data_sliced2 = data.loc[[10,20,30]]
display(data_sliced2)
# Selectionner sur base de l'index et des noms de colonnes
data_sliced3 = data.loc[[10,20,30],['age','chol']]
display(data_sliced3)
data_sliced4 = data[['age','chol']].loc[1:5]
display(data_sliced4)
print(data.age)

display(data_sliced3.dtypes)
data_sliced4.dtypes

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
10   54    1   0     140   239    0        1       160     0      1.2    2   

    caa  thall  output  
10    0      2       1

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
10   54    1   0     140   239    0        1       160     0      1.2    2   
20   59    1   0     135   234    0        1       161     0      0.5    1   
30   41    0   1     105   198    0        1       168     0      0.0    2   

    caa  thall  output  
10    0      2       1  
20    0      3       1  
30    1      2       1

age  chol
10   54   239
20   59   234
30   41   198

age  chol
1   37   250
2   41   204
3   56   236
4   57   354
5   57   192

0      63
1      37
2      41
3      56
4      57
       ..
298    57
299    45
300    68
301    57
302    57
Name: age, Length: 303, dtype: int64


age     int64
chol    int64
dtype: object

age     int64
chol    int64
dtype: object

In [33]:
data[1:10]

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   
5   57    1   0     140   192    0        1       148     0      0.4    1   
6   56    0   1     140   294    0        0       153     0      1.3    1   
7   44    1   1     120   263    0        1       173     0      0.0    2   
8   52    1   2     172   199    1        1       162     0      0.5    2   
9   57    1   2     150   168    0        1       174     0      1.6    2   

   caa  thall  output  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

In [22]:
data.loc[:0,'age':'cp']

age  sex  cp
0   63    1   3

In [103]:
# En sélectionnant un range
data_sliced = data.loc[74:80]
data_sliced

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
74   43    0   2     122   213    0        1       165     0      0.2    1   
75   55    0   1     135   250    0        0       161     0      1.4    1   
76   51    1   2     125   245    1        0       166     0      2.4    1   
77   59    1   1     140   221    0        1       164     1      0.0    2   
78   52    1   1     128   205    1        1       184     0      0.0    2   
79   58    1   2     105   240    0        0       154     1      0.6    1   
80   41    1   2     112   250    0        1       179     0      0.0    2   

    caa  thall  output  
74    0      2       1  
75    0      2       1  
76    0      2       1  
77    0      2       1  
78    0      2       1  
79    0      3       1  
80    0      2       1

In [24]:
data.iloc[5:8,[1,5]]

sex  fbs
5    1    0
6    0    0
7    1    0

In [30]:
data.at[0,'age']

np.int64(63)

In [29]:
data.iat[0,0]

np.int64(63)

In [106]:
# Il est également possible de ne sélectionner qu'une colonne en particulier
# Avec les crochets et la colonne voulue
data_sliced = data[['age','chol']]
data_sliced
data_chol = data['chol']
print (data_sliced)
data_chol

     age  chol
0     63   233
1     37   250
2     41   204
3     56   236
4     57   354
..   ...   ...
298   57   241
299   45   264
300   68   193
301   57   131
302   57   236

[303 rows x 2 columns]


0      233
1      250
2      204
3      236
4      354
      ... 
298    241
299    264
300    193
301    131
302    236
Name: chol, Length: 303, dtype: int64

In [107]:
# Avec plusieurs colonnes
# A noter les doubles crochets
data_sliced = data[['age','output']].loc[54:60]
data_sliced

age  output
54   63       1
55   52       1
56   48       1
57   45       1
58   34       1
59   57       1
60   71       1

**2.2. Boolean indexing**

Nous avons vu jusqu'ici comment sélectionner des parties de notre DataFrame avec les indexes ou le nom des colonnes.

Nous pourrions également faire une sélection basée sur des conditions. 

C'est ce que l'on appelle le **boolean indexing**.

In [ ]:
data['age']>58

age
0     True
1    False
2    False
3    False
4    False
..     ...
298  False
299  False
300   True
301  False
302  False

[303 rows x 1 columns]

In [36]:
data.loc[data['age']>58]

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
13    64    1   3     110   211    0        0       144     1      1.8    1   
17    66    0   3     150   226    0        1       114     0      2.6    0   
19    69    0   3     140   239    0        1       151     0      1.8    2   
20    59    1   0     135   234    0        1       161     0      0.5    1   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
293   67    1   2     152   212    0        0       150     0      0.8    1   
295   63    1   0     140   187    0        0       144     1      4.0    2   
296   63    0   0     124   197    0        1       136     1      0.0    1   
297   59    1   0     164   176    1        0        90     0      1.0    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   

     caa  thall  output  
0      0      1       1  
13     0      2       1  
17     0      2       1  
19     2      2       1  
20     0      3       1  
..   ...    ...     ...  
293    0      3       0  
295    2      3       0  
296    0      2       0  
297    2      1       0  
300    2      3       0  

[104 rows x 14 columns]

In [49]:
# On ne sélectionne que les personnes plus agées que 58 ans
data_sliced = data[data['age'] > 58]
data_sliced
data_sliced2= data[data['age']>58].loc[0:50]
display(data_sliced2)
print(data_sliced2.loc[38])
print(data_sliced2.iloc[10])

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63    1   3     145   233    1        0       150     0      2.3    0   
13   64    1   3     110   211    0        0       144     1      1.8    1   
17   66    0   3     150   226    0        1       114     0      2.6    0   
19   69    0   3     140   239    0        1       151     0      1.8    2   
20   59    1   0     135   234    0        1       161     0      0.5    1   
23   61    1   2     150   243    1        1       137     1      1.0    1   
25   71    0   1     160   302    0        1       162     0      0.4    2   
26   59    1   2     150   212    1        1       157     0      1.6    2   
28   65    0   2     140   417    1        0       157     0      0.8    2   
31   65    1   0     120   177    0        1       140     0      0.4    2   
38   65    0   2     155   269    0        1       148     0      0.8    2   
39   65    0   2     160   360    0        0       151     0      0.8    2   

    caa  thall  output  
0     0      1       1  
13    0      2       1  
17    0      2       1  
19    2      2       1  
20    0      3       1  
23    0      2       1  
25    2      2       1  
26    0      2       1  
28    1      2       1  
31    0      3       1  
38    0      2       1  
39    0      2       1

age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64
age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64


On peut également mettre plusieurs conditions.

**Attention:** les opérateurs logiques "and" et "or" classiques de Python ne fonctionnent pas.

Il faut impérativement utiliser les opérateurs **&** , **~** (not) et **|** et mettre des **()** autour des conditions.

In [23]:
data_sliced = data[(data['age'] > 58) & (data['output'] == 1)]
data_sliced
data_sliced2 = data[(~~data['age'] > 58) & (data['output'] == 1)]
data_sliced2.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63    1   3     145   233    1        0       150     0      2.3    0   
13   64    1   3     110   211    0        0       144     1      1.8    1   
17   66    0   3     150   226    0        1       114     0      2.6    0   
19   69    0   3     140   239    0        1       151     0      1.8    2   
20   59    1   0     135   234    0        1       161     0      0.5    1   

    caa  thall  output  
0     0      1       1  
13    0      2       1  
17    0      2       1  
19    2      2       1  
20    0      3       1

Il est possible de créer de nouvelles colonnes de manière à enrichir vos données.

In [48]:
# Ici on souhaite créer une colonne basée sur la colonne age de manière à obtenir des catégories
# On l'instancie d'abord à 'unknown'
data['age_cat'] = np.nan
display(data.head())
del data['age_cat']

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  age_cat  
0    0      1       1      NaN  
1    0      2       1      NaN  
2    0      2       1      NaN  
3    0      2       1      NaN  
4    0      2       1      NaN

In [ ]:
data['newcol'] = data['age']*data['oldpeak']
display(data.head())
data.drop(columns=['newcol'],inplace=True)
data.head()


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  newcol  
0    0      1       1   144.9  
1    0      2       1   129.5  
2    0      2       1    57.4  
3    0      2       1    44.8  
4    0      2       1    34.2

None

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [31]:
# Nous allons utiliser le boolean indexing pour modifier la catégorie
# Attention à utiliser .loc lorsque l'on chercher un modifier des valeurs
data.loc[data['age'] <= 18, 'age_cat'] = 'child'
display(data)
data.loc[(data['age'] > 18) & (data['age'] < 66), 'age_cat'] = 'adult'
data.loc[data['age'] > 65, 'age_cat'] = 'senior'
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1     NaN  
1      0      2       1     NaN  
2      0      2       1     NaN  
3      0      2       1     NaN  
4      0      2       1     NaN  
..   ...    ...     ...     ...  
298    0      3       0     NaN  
299    0      3       0     NaN  
300    2      3       0     NaN  
301    1      3       0     NaN  
302    1      2       0     NaN  

[303 rows x 15 columns]

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

## <u>3. Opérations mathématiques</u>
- Pandas supporte bien évidemment toutes les opérations mathématiques de base
- Pandas supporte également toutes les opérations de numpy

In [71]:
data_math = cp.copy(data)
# Addition d'une constante à l'entièreté d'une colonne
data_math['age'] += 100
# identique a data_math['age'] = data_math['age'] +100
data_math

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    163    1   3     145   233    1        0       150     0      2.3    0   
1    137    1   2     130   250    0        1       187     0      3.5    0   
2    141    0   1     130   204    0        0       172     0      1.4    2   
3    156    1   1     120   236    0        1       178     0      0.8    2   
4    157    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298  157    0   0     140   241    0        1       123     1      0.2    1   
299  145    1   3     110   264    0        1       132     0      1.2    1   
300  168    1   0     144   193    1        1       141     0      3.4    1   
301  157    1   0     130   131    0        1       115     1      1.2    1   
302  157    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

In [ ]:
# On peut également additionner les colonnes entre elles
data_math['age'] = data_math['age']+data_math['chol']

age           int64
sex           int64
cp            int64
trtbps        int64
chol          int64
fbs           int64
restecg       int64
thalachh      int64
exng          int64
oldpeak     float64
slp           int64
caa           int64
thall         int64
output        int64
dtype: object

On peut retrouver toutes les fonctions mathématiques de numpy ici:

https://numpy.org/doc/stable/reference/routines.math.html

In [ ]:
data_math['age'] = np.log(data_math['age'])
data_math
data_test = np.sum(data_math['age'])
data_test

np.float64(1526.387480233499)

## <u>4. Agrégations et statistiques</u>
Tout comme avec numpy, il est possible d'agréger nos données selon un axe spécifique.

Pandas fournit également quelque statistiques de base.

In [32]:
# Par défaut axis=0
data.sum(axis=0)

age                                                     16473
sex                                                       207
cp                                                        293
trtbps                                                  39882
chol                                                    74618
fbs                                                        45
restecg                                                   160
thalachh                                                45343
exng                                                       99
oldpeak                                                 315.0
slp                                                       424
caa                                                       221
thall                                                     701
output                                                    165
age_cat     adultadultadultadultadultadultadultadultadulta...
dtype: object

In [16]:
data.sum(axis=1, numeric_only=True)  # numeric_only=True

NameError: name 'data' is not defined

In [76]:
display(data.shape)
display(data.ndim)
display(data.columns)
display(data.size)
display(data.dtypes)
display(data.sex)
data.info()

(303, 15)

2

Index(['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output', 'age_cat'],
      dtype='object')

4545

age           int64
sex           int64
cp            int64
trtbps        int64
chol          int64
fbs           int64
restecg       int64
thalachh      int64
exng          int64
oldpeak     float64
slp           int64
caa           int64
thall         int64
output        int64
age_cat      object
dtype: object

0      1
1      1
2      0
3      1
4      0
      ..
298    0
299    1
300    1
301    1
302    0
Name: sex, Length: 303, dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
 14  age_cat   303 non-null    object 
dtypes: float64(1), int64(13), object(1)
memory usage: 35.6+ KB


In [44]:
# Imaginons que nous devions calculer la moyenne pour chaque colonne
print(data.shape)
print(data.sum(numeric_only=True)/data.shape[0])
print(data.sum(axis=1,numeric_only=True)/data.shape[1])

(303, 15)
age          54.366337
sex           0.683168
cp            0.966997
trtbps      131.623762
chol        246.264026
fbs           0.148515
restecg       0.528053
thalachh    149.646865
exng          0.326733
oldpeak       1.039604
slp           1.399340
caa           0.729373
thall         2.313531
output        0.544554
dtype: float64
0      40.020000
1      40.966667
2      36.960000
3      39.920000
4      46.773333
         ...    
298    37.813333
299    37.413333
300    37.226667
301    29.480000
302    40.133333
Length: 303, dtype: float64


In [ ]:
# Plus simple comme ceci:
print(data[['age','chol']].mean())
data[['age','chol']].loc[1:20].mean()

age      54.366337
chol    246.264026
dtype: float64


age      53.75
chol    246.95
dtype: float64

In [81]:
# Lien de corrélation
data.corr(numeric_only=True)

age       sex        cp    trtbps      chol       fbs  \
age       1.000000 -0.098447 -0.068653  0.279351  0.213678  0.121308   
sex      -0.098447  1.000000 -0.049353 -0.056769 -0.197912  0.045032   
cp       -0.068653 -0.049353  1.000000  0.047608 -0.076904  0.094444   
trtbps    0.279351 -0.056769  0.047608  1.000000  0.123174  0.177531   
chol      0.213678 -0.197912 -0.076904  0.123174  1.000000  0.013294   
fbs       0.121308  0.045032  0.094444  0.177531  0.013294  1.000000   
restecg  -0.116211 -0.058196  0.044421 -0.114103 -0.151040 -0.084189   
thalachh -0.398522 -0.044020  0.295762 -0.046698 -0.009940 -0.008567   
exng      0.096801  0.141664 -0.394280  0.067616  0.067023  0.025665   
oldpeak   0.210013  0.096093 -0.149230  0.193216  0.053952  0.005747   
slp      -0.168814 -0.030711  0.119717 -0.121475 -0.004038 -0.059894   
caa       0.276326  0.118261 -0.181053  0.101389  0.070511  0.137979   
thall     0.068001  0.210041 -0.161736  0.062210  0.098803 -0.032019   
output   -0.225439 -0.280937  0.433798 -0.144931 -0.085239 -0.028046   

           restecg  thalachh      exng   oldpeak       slp       caa  \
age      -0.116211 -0.398522  0.096801  0.210013 -0.168814  0.276326   
sex      -0.058196 -0.044020  0.141664  0.096093 -0.030711  0.118261   
cp        0.044421  0.295762 -0.394280 -0.149230  0.119717 -0.181053   
trtbps   -0.114103 -0.046698  0.067616  0.193216 -0.121475  0.101389   
chol     -0.151040 -0.009940  0.067023  0.053952 -0.004038  0.070511   
fbs      -0.084189 -0.008567  0.025665  0.005747 -0.059894  0.137979   
restecg   1.000000  0.044123 -0.070733 -0.058770  0.093045 -0.072042   
thalachh  0.044123  1.000000 -0.378812 -0.344187  0.386784 -0.213177   
exng     -0.070733 -0.378812  1.000000  0.288223 -0.257748  0.115739   
oldpeak  -0.058770 -0.344187  0.288223  1.000000 -0.577537  0.222682   
slp       0.093045  0.386784 -0.257748 -0.577537  1.000000 -0.080155   
caa      -0.072042 -0.213177  0.115739  0.222682 -0.080155  1.000000   
thall    -0.011981 -0.096439  0.206754  0.210244 -0.104764  0.151832   
output    0.137230  0.421741 -0.436757 -0.430696  0.345877 -0.391724   

             thall    output  
age       0.068001 -0.225439  
sex       0.210041 -0.280937  
cp       -0.161736  0.433798  
trtbps    0.062210 -0.144931  
chol      0.098803 -0.085239  
fbs      -0.032019 -0.028046  
restecg  -0.011981  0.137230  
thalachh -0.096439  0.421741  
exng      0.206754 -0.436757  
oldpeak   0.210244 -0.430696  
slp      -0.104764  0.345877  
caa       0.151832 -0.391724  
thall     1.000000 -0.344029  
output   -0.344029  1.000000

Nous avons vu comment créer des catégories. Comment calculer la moyenne par exemple en fonction de ces catégories?

Avec la fonction **groupby** comme en SQL.


In [41]:
data.groupby(['age_cat','output'],as_index=False).sum()

age_cat  output   age  sex   cp  trtbps   chol  fbs  restecg  thalachh  \
0   adult       0  6719   99   57   16296  30657   18       55     17088   
1   adult       1  7490   88  201   19050  35482   21       89     23773   
2  senior       0  1092   15    9    2251   3993    4        7      2108   
3  senior       1  1172    5   26    2285   4486    2        9      2374   

   exng  oldpeak  slp  caa  thall  
0    66    195.8  144  134    311  
1    22     81.3  238   48    315  
2    10     23.0   17   27     40  
3     1     14.9   25   12     35

In [89]:
# Il suffit de spécifier la colonne
data.loc[data['age'] <= 18, 'age_cat'] = 'child'
data.loc[(data['age'] > 18) & (data['age'] < 66), 'age_cat'] = 'adult'
data.loc[data['age'] > 65, 'age_cat'] = 'senior'
grouped_data = data.groupby(['age_cat','output']).mean(numeric_only=True)
display(grouped_data)
grouped_data.reset_index()

age       sex        cp      trtbps        chol  \
age_cat output                                                          
adult   0       55.073770  0.811475  0.467213  133.573770  251.286885   
        1       50.608108  0.594595  1.358108  128.716216  239.743243   
senior  0       68.250000  0.937500  0.562500  140.687500  249.562500   
        1       68.941176  0.294118  1.529412  134.411765  263.882353   

                     fbs   restecg    thalachh      exng   oldpeak       slp  \
age_cat output                                                                 
adult   0       0.147541  0.450820  140.065574  0.540984  1.604918  1.180328   
        1       0.141892  0.601351  160.628378  0.148649  0.549324  1.608108   
senior  0       0.250000  0.437500  131.750000  0.625000  1.437500  1.062500   
        1       0.117647  0.529412  139.647059  0.058824  0.876471  1.470588   

                     caa     thall  
age_cat output                      
adult   0       1.098361  2.549180  
        1       0.324324  2.128378  
senior  0       1.687500  2.500000  
        1       0.705882  2.058824

age_cat  output        age       sex        cp      trtbps        chol  \
0   adult       0  55.073770  0.811475  0.467213  133.573770  251.286885   
1   adult       1  50.608108  0.594595  1.358108  128.716216  239.743243   
2  senior       0  68.250000  0.937500  0.562500  140.687500  249.562500   
3  senior       1  68.941176  0.294118  1.529412  134.411765  263.882353   

        fbs   restecg    thalachh      exng   oldpeak       slp       caa  \
0  0.147541  0.450820  140.065574  0.540984  1.604918  1.180328  1.098361   
1  0.141892  0.601351  160.628378  0.148649  0.549324  1.608108  0.324324   
2  0.250000  0.437500  131.750000  0.625000  1.437500  1.062500  1.687500   
3  0.117647  0.529412  139.647059  0.058824  0.876471  1.470588  0.705882   

      thall  
0  2.549180  
1  2.128378  
2  2.500000  
3  2.058824

In [85]:
grouped_data.index

MultiIndex([( 'adult', 0),
            ( 'adult', 1),
            ('senior', 0),
            ('senior', 1)],
           names=['age_cat', 'output'])

In [53]:
data.groupby(['age_cat','output']).agg({'age':['mean','max','min'],
                                        'chol':'max'})

age         chol
                     mean max min  max
age_cat output                        
adult   0       55.073770  65  35  409
        1       50.608108  65  29  417
senior  0       68.250000  77  66  322
        1       68.941176  76  66  564

In [93]:
# Nous n'avons pas la catégorie child, pourquoi?
# Value counts permet d'inventorier nos valeurs catégorielles
data['age_cat'].value_counts()

age_cat
adult     270
senior     33
Name: count, dtype: int64

In [54]:
# Vérifions
print(data['age'].min())
data['age'].max()


29


np.int64(77)

In [44]:
# La fonction describe permet d'obtenir une vision très macro de vos données
data.describe(include='all').T
data.groupby(by=['age-cat']).describe()

count unique    top freq        mean        std    min    25%  \
age       303.0    NaN    NaN  NaN   54.366337   9.082101   29.0   47.5   
sex       303.0    NaN    NaN  NaN    0.683168   0.466011    0.0    0.0   
cp        303.0    NaN    NaN  NaN    0.966997   1.032052    0.0    0.0   
trtbps    303.0    NaN    NaN  NaN  131.623762  17.538143   94.0  120.0   
chol      303.0    NaN    NaN  NaN  246.264026  51.830751  126.0  211.0   
fbs       303.0    NaN    NaN  NaN    0.148515   0.356198    0.0    0.0   
restecg   303.0    NaN    NaN  NaN    0.528053    0.52586    0.0    0.0   
thalachh  303.0    NaN    NaN  NaN  149.646865  22.905161   71.0  133.5   
exng      303.0    NaN    NaN  NaN    0.326733   0.469794    0.0    0.0   
oldpeak   303.0    NaN    NaN  NaN    1.039604   1.161075    0.0    0.0   
slp       303.0    NaN    NaN  NaN     1.39934   0.616226    0.0    1.0   
caa       303.0    NaN    NaN  NaN    0.729373   1.022606    0.0    0.0   
thall     303.0    NaN    NaN  NaN    2.313531   0.612277    0.0    2.0   
output    303.0    NaN    NaN  NaN    0.544554   0.498835    0.0    0.0   
age_cat     303      2  adult  270         NaN        NaN    NaN    NaN   

            50%    75%    max  
age        55.0   61.0   77.0  
sex         1.0    1.0    1.0  
cp          1.0    2.0    3.0  
trtbps    130.0  140.0  200.0  
chol      240.0  274.5  564.0  
fbs         0.0    0.0    1.0  
restecg     1.0    1.0    2.0  
thalachh  153.0  166.0  202.0  
exng        0.0    1.0    1.0  
oldpeak     0.8    1.6    6.2  
slp         1.0    2.0    2.0  
caa         0.0    1.0    4.0  
thall       2.0    3.0    3.0  
output      1.0    1.0    1.0  
age_cat     NaN    NaN    NaN

In [118]:
print(data.shape)
print(50*'-')
print(data.size)
print(50*'-')
print(data.dtypes)
print(50*'-')
print(data.columns)
print(50*'-')
print(data.index)
print(50*'-')
print(data.values)

(303, 15)
--------------------------------------------------
4545
--------------------------------------------------
age           int64
sex           int64
cp            int64
trtbps        int64
chol          int64
fbs           int64
restecg       int64
thalachh      int64
exng          int64
oldpeak     float64
slp           int64
caa           int64
thall         int64
output        int64
age_cat      object
dtype: object
--------------------------------------------------
Index(['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output', 'age_cat'],
      dtype='object')
--------------------------------------------------
RangeIndex(start=0, stop=303, step=1)
--------------------------------------------------
[[63 1 3 ... 1 1 'adult']
 [37 1 2 ... 2 1 'adult']
 [41 0 1 ... 2 1 'adult']
 ...
 [68 1 0 ... 3 0 'senior']
 [57 1 0 ... 3 0 'adult']
 [57 0 1 ... 2 0 'adult']]


In [119]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
 14  age_cat   303 non-null    object 
dtypes: float64(1), int64(13), object(1)
memory usage: 35.6+ KB


## <u>5. Comment gérer les valeurs manquantes:</u>

In [45]:
# D'abord modifions notre dataset de sorte à avoir des données manquantes
# nan correspond à Null, aux valeurs manquantes
# Ici l'âge pour toutes les personnes de moins de 58 ans sera à NaN
data.loc[data['age'] < 58, 'age'] = np.nan
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63.0    1   3     145   233    1        0       150     0      2.3    0   
1     NaN    1   2     130   250    0        1       187     0      3.5    0   
2     NaN    0   1     130   204    0        0       172     0      1.4    2   
3     NaN    1   1     120   236    0        1       178     0      0.8    2   
4     NaN    0   0     120   354    0        1       163     1      0.6    2   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   NaN    0   0     140   241    0        1       123     1      0.2    1   
299   NaN    1   3     110   264    0        1       132     0      1.2    1   
300  68.0    1   0     144   193    1        1       141     0      3.4    1   
301   NaN    1   0     130   131    0        1       115     1      1.2    1   
302   NaN    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

In [49]:
# D'abord il faut les identifier
# isna() permet de renvoyer True ou False si les données sont manquantes ou non
# Ensuite sum() additionne le tout. False étant == 0 et True == 1 on peut en déduire le nombre d'élément manquant
data.isna().sum()

age         180
sex           0
cp            0
trtbps        0
chol          0
fbs           0
restecg       0
thalachh      0
exng          0
oldpeak       0
slp           0
caa           0
thall         0
output        0
age_cat       0
dtype: int64

In [ ]:
# fillna() permet de remplacer des valeurs manquantes par une valeur choisie
# ici on choisir de remplacer par la moyenne
data['age'] = data['age'].fillna(data['age'].mean())
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63.0    1   3     145   233    1        0       150     0      2.3    0   
1    63.0    1   2     130   250    0        1       187     0      3.5    0   
2    63.0    0   1     130   204    0        0       172     0      1.4    2   
3    63.0    1   1     120   236    0        1       178     0      0.8    2   
4    63.0    0   0     120   354    0        1       163     1      0.6    2   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298  63.0    0   0     140   241    0        1       123     1      0.2    1   
299  63.0    1   3     110   264    0        1       132     0      1.2    1   
300  68.0    1   0     144   193    1        1       141     0      3.4    1   
301  63.0    1   0     130   131    0        1       115     1      1.2    1   
302  63.0    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

On pourrait également supprimer les valeurs manquantes.

Attention à ne pas trop supprimer de lignes ou de colonnes.

In [66]:
data['age'].sort_values()
data.sort_values(by=['age','chol'],ascending=[False,True])

16     58.0
14     58.0
79     58.0
64     58.0
89     58.0
       ... 
294     NaN
298     NaN
299     NaN
301     NaN
302     NaN
Name: age, Length: 303, dtype: float64

In [68]:
data.loc[data['age'] == 63, 'age'] = np.nan
data = data.dropna(subset=['age'])
data.sort_values(by=['age','chol'],ascending=[True,False])

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
16   58.0    0   2     120   340    0        1       172     0      0.0    2   
278  58.0    0   1     136   319    1        0       152     0      0.0    2   
291  58.0    1   0     114   318    0        2       140     0      4.4    0   
236  58.0    1   0     125   300    0        0       171     0      0.0    2   
172  58.0    1   1     120   284    0        0       160     0      1.8    1   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
60   71.0    0   2     110   265    1        0       130     0      0.0    2   
151  71.0    0   0     112   149    0        1       125     0      1.6    1   
129  74.0    0   1     120   269    0        0       121     1      0.2    2   
144  76.0    0   2     140   197    0        2       116     0      1.1    1   
238  77.0    1   0     125   304    0        0       162     1      0.0    2   

     caa  thall  output age_cat  
16     0      2       1   adult  
278    2      2       0   adult  
291    3      1       0   adult  
236    2      3       0   adult  
172    0      2       0   adult  
..   ...    ...     ...     ...  
60     1      2       1  senior  
151    0      2       1  senior  
129    1      2       1  senior  
144    0      2       1  senior  
238    3      2       0  senior  

[114 rows x 15 columns]

Nous sommes passés de 303 lignes à 114, ce qui fait beaucoup.

Il serait plus sage de supprimer la colonne age plutôt que de supprimer chaque ligne où il y a une valeur manquante.

## <u>6. Comment join plusieurs datasets </U>

Fonctionnement identique au join SQL

Illustation avec le pd.read_clipboard()

PIB: https://fr.wikipedia.org/wiki/Produit_int%C3%A9rieur_brut

Population: https://fr.wikipedia.org/wiki/Liste_des_pays_par_population

In [20]:
pib = pd.read_clipboard()
pib

Pays   2014   2015   2016   \
0                   Drapeau du Luxembourg Luxembourg     283    282    278   
1                  Drapeau des États-Unis États-Unis     148    148    142   
2                    Drapeau de la Belgique Belgique     121    121    120   
3                   Drapeau de l'Allemagne Allemagne     127    124    125   
4                        Drapeau de la France France     108    107    106   
5   Drapeau de l’Union européenne Union européenne...    100    100    100   
6                 Drapeau du Royaume-Uni Royaume-Uni     111    111    109   
7                         Drapeau de l'Italie Italie      98     97     99   
8                       Drapeau de l'Espagne Espagne      90     91     92   
9                             Drapeau du Japon Japon     107    107    100   
10                     Drapeau de la Pologne Pologne      67     69     69   
11                     Drapeau de la Turquie Turquie      65     68     66   

    2017   2018   2019   2020   2021   2022   2023  
0     269    260    252    259    266    256   239  
1     140    140    134    135    135    134   136  
2     118    118    118    118    120    120   118  
3     124    124    121    123    120    117   115  
4     104    104    106    105    103    100   101  
5     100    100    100    100    100    100   100  
6     108    105    104    102    102     99    98  
7      98     97     97     94     96     98    97  
8      93     91     91     83     84     85    88  
9      97     95     89     90     86     82    84  
10     69     71     73     76     77     80    80  
11     66     63     59     61     61     67    73

In [51]:

pib['Pays'] = pib['Pays'].apply(lambda x: x.split()[-1])
pib

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023
0    Luxembourg   283   282   278   269   260   252   259   266   256   239
1    États-Unis   148   148   142   140   140   134   135   135   134   136
2      Belgique   121   121   120   118   118   118   118   120   120   118
3     Allemagne   127   124   125   124   124   121   123   120   117   115
4        France   108   107   106   104   104   106   105   103   100   101
5         pays)   100   100   100   100   100   100   100   100   100   100
6   Royaume-Uni   111   111   109   108   105   104   102   102    99    98
7        Italie    98    97    99    98    97    97    94    96    98    97
8       Espagne    90    91    92    93    91    91    83    84    85    88
9         Japon   107   107   100    97    95    89    90    86    82    84
10      Pologne    67    69    69    69    71    73    76    77    80    80
11      Turquie    65    68    66    66    63    59    61    61    67    73

In [52]:
pib = pib.drop(5)


In [73]:
pib['Pays'].unique()

array(['Luxembourg', 'États-Unis', 'Belgique', 'Allemagne', 'France',
       'Royaume-Uni', 'Italie', 'Espagne', 'Japon', 'Pologne', 'Turquie'],
      dtype=object)

In [25]:
pib

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023
0    Luxembourg   283   282   278   269   260   252   259   266   256   239
1    États-Unis   148   148   142   140   140   134   135   135   134   136
2      Belgique   121   121   120   118   118   118   118   120   120   118
3     Allemagne   127   124   125   124   124   121   123   120   117   115
4        France   108   107   106   104   104   106   105   103   100   101
6   Royaume-Uni   111   111   109   108   105   104   102   102    99    98
7        Italie    98    97    99    98    97    97    94    96    98    97
8       Espagne    90    91    92    93    91    91    83    84    85    88
9         Japon   107   107   100    97    95    89    90    86    82    84
10      Pologne    67    69    69    69    71    73    76    77    80    80
11      Turquie    65    68    66    66    63    59    61    61    67    73

In [53]:
pib.melt(id_vars=['Pays'],value_vars=pib.drop(columns=['Pays']).columns,var_name='Annee')

Pays Annee  value
0    Luxembourg  2014    283
1    États-Unis  2014    148
2      Belgique  2014    121
3     Allemagne  2014    127
4        France  2014    108
..          ...   ...    ...
105      Italie  2023     97
106     Espagne  2023     88
107       Japon  2023     84
108     Pologne  2023     80
109     Turquie  2023     73

[110 rows x 3 columns]

In [21]:
url = 'https://fr.wikipedia.org/wiki/Liste_des_pays_par_population'
response = rq.get(url,headers={'User-Agent': 'Mozilla/5.0'})
data = pd.read_html(StringIO(response.text))
population = data[0]
population

Rang (2023) Pays ou territoire Population au 1er juillet 2023  \
0             -              Monde                  8 091 734 930   
1             1            Inde[a]                  1 438 069 596   
2             2           Chine[b]                  1 422 584 933   
3             3      États-Unis[c]                    343 477 335   
4             4          Indonésie                    281 190 067   
..          ...                ...                            ...   
233           -         Montserrat                          4 420   
234           -     Îles Malouines                          3 477   
235           -            Tokelau                          2 397   
236         199               Niue                          1 817   
237         200            Vatican                            496   

    Population projetée au 1er janvier 2026  
0                             8 266 245 291  
1                             1 470 295 711  
2                             1 414 458 552  
3                               348 178 045  
4                               286 820 254  
..                                      ...  
233                                   4 346  
234                                   3 467  
235                                   2 656  
236                                   1 819  
237                                     503  

[238 rows x 4 columns]

In [59]:
population['Pays ou territoire'] = population['Pays ou territoire'].str.replace(r'\[.*\]', '', regex=True).str.strip()

In [22]:
population

Rang (2023) Pays ou territoire Population au 1er juillet 2023  \
0             -              Monde                  8 091 734 930   
1             1            Inde[a]                  1 438 069 596   
2             2           Chine[b]                  1 422 584 933   
3             3      États-Unis[c]                    343 477 335   
4             4          Indonésie                    281 190 067   
..          ...                ...                            ...   
233           -         Montserrat                          4 420   
234           -     Îles Malouines                          3 477   
235           -            Tokelau                          2 397   
236         199               Niue                          1 817   
237         200            Vatican                            496   

    Population projetée au 1er janvier 2026  
0                             8 266 245 291  
1                             1 470 295 711  
2                             1 414 458 552  
3                               348 178 045  
4                               286 820 254  
..                                      ...  
233                                   4 346  
234                                   3 467  
235                                   2 656  
236                                   1 819  
237                                     503  

[238 rows x 4 columns]

In [61]:
print(list(population))
print(list(pib))
finaldf = pib.merge(population, how='left', left_on='Pays', right_on='Pays ou territoire')
finaldf


['Rang (2023)', 'Pays ou territoire', 'Population au 1er juillet 2023', 'Population projetée au 1er juillet 2025']
['Pays', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023']


Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  \
0    Luxembourg   283   282   278   269   260   252   259   266   256   239   
1    États-Unis   148   148   142   140   140   134   135   135   134   136   
2      Belgique   121   121   120   118   118   118   118   120   120   118   
3     Allemagne   127   124   125   124   124   121   123   120   117   115   
4        France   108   107   106   104   104   106   105   103   100   101   
5   Royaume-Uni   111   111   109   108   105   104   102   102    99    98   
6        Italie    98    97    99    98    97    97    94    96    98    97   
7       Espagne    90    91    92    93    91    91    83    84    85    88   
8         Japon   107   107   100    97    95    89    90    86    82    84   
9       Pologne    67    69    69    69    71    73    76    77    80    80   
10      Turquie    65    68    66    66    63    59    61    61    67    73   

   Rang (2023) Pays ou territoire Population au 1er juillet 2023  \
0          165         Luxembourg                        665 098   
1            3         États-Unis                    343 477 335   
2           81           Belgique                     11 712 893   
3           19          Allemagne                     84 548 231   
4           23             France                     66 438 822   
5           21        Royaume-Uni                     68 682 962   
6           25             Italie                     59 499 453   
7           32            Espagne                     47 911 579   
8           12              Japon                    124 370 947   
9           39            Pologne                     38 762 844   
10          18            Turquie                     87 270 501   

   Population projetée au 1er juillet 2025  
0                                  680 453  
1                              347 275 807  
2                               11 758 603  
3                               84 075 075  
4                               66 650 804  
5                               69 551 332  
6                               59 146 260  
7                               47 889 958  
8                              123 103 479  
9                               38 140 910  
10                              87 685 426

In [90]:
finaldf = finaldf.rename(columns={'Rang (2023)':'Rang population'})
finaldf

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  \
0    Luxembourg   283   282   278   269   260   252   259   266   256   239   
1    États-Unis   148   148   142   140   140   134   135   135   134   136   
2      Belgique   121   121   120   118   118   118   118   120   120   118   
3     Allemagne   127   124   125   124   124   121   123   120   117   115   
4        France   108   107   106   104   104   106   105   103   100   101   
5   Royaume-Uni   111   111   109   108   105   104   102   102    99    98   
6        Italie    98    97    99    98    97    97    94    96    98    97   
7       Espagne    90    91    92    93    91    91    83    84    85    88   
8         Japon   107   107   100    97    95    89    90    86    82    84   
9       Pologne    67    69    69    69    71    73    76    77    80    80   
10      Turquie    65    68    66    66    63    59    61    61    67    73   

   Rang population Pays ou territoire Population au 1er juillet 2023  \
0              165         Luxembourg                        665 098   
1                3         États-Unis                    343 477 335   
2               81           Belgique                     11 712 893   
3               19          Allemagne                     84 548 231   
4               23             France                     66 438 822   
5               21        Royaume-Uni                     68 682 962   
6               25             Italie                     59 499 453   
7               32            Espagne                     47 911 579   
8               12              Japon                    124 370 947   
9               39            Pologne                     38 762 844   
10              18            Turquie                     87 270 501   

   Population projetée au 1er juillet 2025  
0                                  680 453  
1                              347 275 807  
2                               11 758 603  
3                               84 075 075  
4                               66 650 804  
5                               69 551 332  
6                               59 146 260  
7                               47 889 958  
8                              123 103 479  
9                               38 140 910  
10                              87 685 426

## <u>7. connection db </U>

<br>

Créer une connecteur sur base d'une URL avec comme information:
- le SGBD
- username+password
- adresse réseau
- le nom de la db à laquelle se connecter

E.g.:
- postgresql://username:password@localhost:5432/db_name # localhost=127.0.0.1
- mssql+pyodbc://username:password@localhost:1433/db_name
- mssql+pyodbc://server_name/db_name?trusted_connection=yes # auth windows
- mssql+pyodbc://server_name/db_name?trusted_connection=yes&driver=ODBC+Driver+17+for+SQL+Server # Avec driver ODBC spécifique

In [92]:
sgbd = 'mssql+pyodbc'
server_name = r'GEORGES\DATA'
db_name = 'DBSlide'
driver = 'ODBC+Driver+17+for+SQL+Server'

url = f'{sgbd}://{server_name}/{db_name}?trusted_connection=yes&driver={driver}'
con = create_engine(url)

sql_query = "SELECT * FROM student WHERE year_result > 15"
data = pd.read_sql(sql_query,con)

con.dispose() # release con

data

student_id first_name last_name birth_date     login  section_id  \
0           5      Kevin     Bacon 1958-07-08    kbacon        1120   
1           8      Julia   Roberts 1967-10-28  jroberts        1120   
2          11       Andy    Garcia 1956-04-12   agarcia        1110   
3          18   Jennifer    Garner 1972-04-17   jgarner        1120   
4          25      Halle     Berry 1966-08-14    hberry        1320   

   year_result course_id  
0           16         0  
1           17         0  
2           19         0  
3           18         0  
4           18    EG2210